### PR TITLE
fix: check compressed data compatibility

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Temporary fix, backward compatible.

rollup-relayer will handle this special error and early stop adding more blocks into a chunk (and more chunks into a batch).

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] fix: A bug fix

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

**Review by noel2004 — APPROVED** [review]

File: encoding/codecv2/codecv2.go
Comment (Thegaram) [inline encoding/codecv2/codecv2.go:313]:
```suggestion
			return math.MaxUint32, math.MaxUint32, nil // Return math.MaxUint32 to indicate the chunk is too large and avoid overflows.
```

File: encoding/codecv2/codecv2.go
Comment (Thegaram) [inline encoding/codecv2/codecv2.go:332]: Let's add logs + metrics to these 2 error cases so that we can see how often they happen.

**Review by Thegaram — COMMENTED** [review]
Looks good, let's also wait for @roynalnaruto's review before merging.

File: encoding/codecv2/codecv2.go
Comment (Thegaram) [reply encoding/codecv2/codecv2.go:332]: We can also add these on the caller side (in `chunk-proposer` and `batch-proposer`), but in that case it's better to return an error here. Your call @colinlyguo

File: encoding/da.go
Comment (noel2004) [reply encoding/da.go:297]: fheader is a `byte` (i.e uint8) and there is impossible for a unit8 >> 6 obtain value beyond 0-3.

So it should be safe to panic here (like the unreachable block in a function)

File: encoding/codecv2/codecv2.go
Comment (colinlyguo) [reply encoding/codecv2/codecv2.go:342]: add a new error type (`CompressedDataCompatibilityError`) for `chunk-proposer` and `batch-proposer` to catch, will submit metrics in `chunk-proposer` and `batch-proposer`.

File: encoding/codecv2/codecv2.go
Comment (Thegaram) [inline encoding/codecv2/codecv2.go:313]:
```suggestion
			log.Warn("EstimateChunkL1CommitBatchSizeAndBlobSize: compressed data compatibility check failed", "err", err, "batchBytes", hex.EncodeToString(batchBytes), "blobBytes", hex.EncodeToString(blobBytes))
```

**Review by Thegaram — APPROVED** [review]

File: encoding/codecv2/codecv2.go
Comment (colinlyguo) [reply encoding/codecv2/codecv2.go]: changed in [3c6edd8](https://github.com/scroll-tech/da-codec/pull/17/commits/3c6edd82ce0f68a88a60f7baccb2aee6b5a6cd00).

**Review by Thegaram — APPROVED** [review]

**Review by noel2004 — APPROVED** [review]

File: encoding/codecv2/codecv2.go
Comment (roynalnaruto) [inline encoding/codecv2/codecv2.go:194]: @colinlyguo when does this scenario occur? `blobBytes < 4096 * 31` will always hold, right? So isn't this check redundant?

We should instead have this check for `batchBytes`, which is already done [here](https://github.com/scroll-tech/da-codec/pull/17/files#diff-626a61d73695a9154e13d35f43a9633078eacc8ce3e1cb8a290e8c0723a0a0dcR319-R325)

File: encoding/codecv2/codecv2.go
Comment (roynalnaruto) [inline encoding/codecv2/codecv2.go:188]: This should be `len(batchBytes)` ?

**Review by roynalnaruto — COMMENTED** [review]
May be an issue here?

File: encoding/codecv2/codecv2.go
Comment (colinlyguo) [reply encoding/codecv2/codecv2.go:188]: nice catch, this is fixed in https://github.com/scroll-tech/da-codec/pull/18. when using the new patch in rollup-relayer, some unit tests failed due to the bug you mentioned.

File: encoding/codecv2/codecv2.go
Comment (Thegaram) [reply encoding/codecv2/codecv2.go:194]: `blobBytes` here is uncompressed (as opposed to `compressedBlobBytes` a few lines above).

File: encoding/codecv2/codecv2.go
Comment (colinlyguo) [reply encoding/codecv2/codecv2.go:188]: blobBytes here means batchBytes, the naming is outdated.

**Pull request merged**
